### PR TITLE
Poster generation: Set video as the poster's parent 

### LINKED ISSUE #5544: Poster generation: Set video as the poster's parent
## Feature Description

Splitting this out from #5316.

Would make sense to set the video as the poster's parent to make it clearer to user where this poster is being used.

My first though was to simply add `post: id` here (`id` is the video ID):

https://github.com/google/web-stories-wp/blob/c170896a31487c8c55a234f3f770baf5b5d076f6/assets/src/edit-story/app/media/utils/useUploadVideoFrame.js#L54-L59

However, the REST API does not allow setting the parent to an attachment (or revision):

https://github.com/WordPress/wordpress-develop/blob/f1e610531a4dcdd817753edc48e311cd3f74d4f7/src/wp-includes/rest-api/endpoints/class-wp-rest-attachments-controller.php#L316-L322

So we'd need to add some workaround in `Stories_Media_Controller` to make that work.

Perhaps hooking into `rest_after_insert_attachment` and manually setting the parent if `web_stories_is_poster` meta key is set or something.

## Additional Context

---

_Do not alter or remove anything below. The following sections will be managed by moderators only._

### PR DESCRIPTION
**Pull request by spacedmonkey**
## Summary

Poster generation: Set video as the poster's parent 

## Relevant Technical Choices

Add a new param to set parent of attachment on update. 
## Testing Instructions

---

Fixes #5544

### CODE REVIEW

**Comment by swissspidy** [comment]
Could you elaborate on why a new param was needed? Because of the limitations outlined on the ticket? Just so it‘s documented...

Also, should there be some check that the parent is actually an attachment? Right now it could be anything.

File: includes/REST_API/Stories_Media_Controller.php [hunk 1]
```diff
@@ -69,6 +69,33 @@ public function get_items( $request ) {
...
+				'ID'          => $request['id'],
+				'post_parent' => $request['web_stories_parent'],
+			];
+			$result = wp_update_post( $args, true );
+			if ( is_wp_error( $result ) ) {
+				return $result;
+			}
+		}
+
+		return parent::update_item( $request );
```
Comment (swissspidy) [inline includes/REST_API/Stories_Media_Controller.php]: Should this happen before the above logic? If there was an error in `parent::update_item( $request )` we should not proceed with our custom logic.

Comment (swissspidy) [inline includes/REST_API/Stories_Media_Controller.php]: Could we simply use `$request['post']` instead? Perhaps like this to work around the parent class limitation:

```php
// WP_REST_Attachments_Controller doesn't allow setting an attachment as the parent post.
// Hence we are working around this here.
$parent_post = ! empty( $request['post'] ) ? (int) $request['post'] : null;
unset( $request['post'] );

$response = parent::update_item( $request );

if ( is_wp_error( $response )  ) {
	return $response;
}

// TODO: Also check for `current_user_can( 'edit_post', (int) $parent_post )` ?
if ( $parent_post && 'attachment' === get_post_type( $parent_post ) ) {
	$args   = [
		'ID'          => $request['id'],
		'post_parent' => $parent_post,
	];
	wp_update_post( $args, true );
}

$attachment = get_post( $request['id'] );

// Ensures that the newly updated post parent is part of the response.
$response = $this->prepare_item_for_response( $attachment, $request );
return rest_ensure_response( $response );
```

File: includes/REST_API/Stories_Media_Controller.php [hunk 2]
```diff
@@ -69,6 +69,67 @@ public function get_items( $request ) {
...
+
+		/** This action is documented in wp-includes/rest-api/endpoints/class-wp-rest-attachments-controller.php */
+		do_action( 'rest_after_insert_attachment', $attachment, $request, false );
+
+		/** This action is documented in wp-includes/post.php */
+		do_action( 'wp_after_insert_post', $post_id, $attachment, true, $attachment_before );
+
+		$response = $this->prepare_item_for_response( $attachment, $request );
+
+		return rest_ensure_response( $response );
```
Comment (swissspidy) [inline includes/REST_API/Stories_Media_Controller.php]: I was hoping we could get away with less duplication.

Could you help me understand why this was necessary here?

File: tests/phpunit/tests/REST_API/Stories_Media_Controller.php [hunk 3]
```diff
@@ -152,4 +152,58 @@ public function test_get_items_filter_video() {
...
+				'post_mime_type' => 'image/jpeg',
+				'post_title'     => 'Test Image',
+			]
+		);
+		wp_set_current_user( self::$user_id );
+		$request = new WP_REST_Request( \WP_REST_Server::CREATABLE, '/web-stories/v1/media/' . $poster_attachment_id );
+		$request->set_param( 'post', $video_attachment_id );
+		$response = rest_get_server()->dispatch( $request );
+		$data     = $response->get_data();
+		$this->assertArrayHasKey( 'post', $data );
```
Comment (swissspidy) [inline tests/phpunit/tests/REST_API/Stories_Media_Controller.php:181]: Maybe also add an assertion that the parent post was actually changed?

File: includes/REST_API/Stories_Media_Controller.php (hunk 2 above)
Comment (spacedmonkey) [reply includes/REST_API/Stories_Media_Controller.php]: I am not sure what you mean. We want to maintain the actions here right?

Comment (swissspidy) [reply includes/REST_API/Stories_Media_Controller.php]: Yes, but when I wrote https://github.com/google/web-stories-wp/pull/5573#discussion_r536667098 I was hoping we could do this without duplicating the code from `WP_REST_Attachments_Controller`.

Basically:

```php
// Run parent class code
$response = parent::update_item( $request )

// Manually set parent
wp_update_post( ... )

// Ensure that $response now contains the correct post parent information
$response['post'] = $parent_id;

return $response
```

Comment (spacedmonkey) [reply includes/REST_API/Stories_Media_Controller.php]: Remove some of the repeated logic. https://github.com/google/web-stories-wp/pull/5573/commits/70d61d7cfb6fb9ee315699ec1de1535acbe89dba

The code snippet above would not work as the filter and actions would not have the correct value for the attachment object.

**Review by swissspidy — APPROVED** [review]
Works as expected!

**Pull request merged**